app/src/org/commcare/android/adapters/HomeScreenAdapter.java

```diff
@@ -0,0 +1,186 @@
+        if(!isInitialized){
+            visibleButtons = new LinkedList<SquareButtonWithNotification>();
+            Log.i("HomeScrnAdpt","Creating all buttons because got a null in position " + position);
+            for (int i = 0; i < buttons.length; i++) {
+                if (buttons[i] != null) continue;
+                SquareButtonWithNotification button = (SquareButtonWithNotification) LayoutInflater.from(context)
+                        .inflate(buttonsResources[i], parent, false);
+                buttons[i] = button;
+                Log.i("HomeScrnAdpt","Added button " + button + "to position " + i);
+
+                View.OnClickListener listener = buttonListeners[i];
+                // creating now, but set a clickListener before, so we'll add it to this button...
+                if(listener != null) {
+                    button.setOnClickListener(listener);
+                    Log.i("HomeScrnAdpt","Added onClickListener " + listener + " to button in position " + i);
+                }
+                if( i == position ) view = button;
+                if(!hiddenButtons[i]) visibleButtons.add(button);
+            }
+            isInitialized = true;
+        }
+        if(position < 0 || position >= getCount()) return null;
+        if(convertView != null) {
+            return convertView;
+        } else {
+            SquareButtonWithNotification btn = visibleButtons.get(position);
+
+            if(btn == null) {
+                Log.i("HomeScrnAdpt","Unexpected null button");
+            }
+
+            return btn;
+        }
+    }
+
+    public void setButtonVisibility(int androidCode, boolean lookupID, boolean isButtonHidden){
+        int index = getButtonIndex(androidCode, lookupID);
+        boolean toggled = isButtonHidden ^ hiddenButtons[index];
+        hiddenButtons[index] = isButtonHidden;
+        if (visibleButtons != null) {
+            if(!toggled) return;
+            if(isButtonHidden) {
+                visibleButtons.remove(buttons[index]);
+            } else {
+                visibleButtons.add(index, buttons[index]);
+            }
+        }
+    }
+
+    //endregion
+
+    //region Private methods
+
+    private void setOnClickListenerForButton(int buttonIndex, View.OnClickListener listener) {
+        buttonListeners[buttonIndex] = listener;
+        SquareButtonWithNotification button = (SquareButtonWithNotification) getItem(buttonIndex);
+        if(button != null){
+            button.setOnClickListener(listener);
+        }
+    }
+
+    /**
+     * Returns the index of the button with the given resource code. If lookupID is set, will search for the button with the given R.id; if not, will search for the button with the given R.layout code.
+     * @param androidCode
+     * @param lookupID
+     * @return
+     * @throws java.lang.IllegalArgumentException If the given androidCode is not found
+     */
+    private int getButtonIndex(int androidCode, boolean lookupID){
```

Comment:
Every current invocation of this function occurs with 'lookupID' set to false. Just want to make sure that all this extra logic is needed (didn't read it closely enough to really understand the mappings)